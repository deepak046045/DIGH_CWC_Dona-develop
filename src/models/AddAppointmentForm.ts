import { AddAppointmentFormFields } from '@/shared/enums/AddAppointmentFormFields';
import { AppointmentTypes } from '@/shared/enums/AppointmentTypes';

export type AddAppointmentForm = {
  [AddAppointmentFormFields.APPOINTMENT_TYPE]: AppointmentTypes | '';
  [AddAppointmentFormFields.INFUSION_NUMBER]: string;
  [AddAppointmentFormFields.ADDRESS]: string;
  [AddAppointmentFormFields.SCHEDULE_DATE]: string;
  [AddAppointmentFormFields.SCHEDULE_TIME]: number;
  [AddAppointmentFormFields.CARE_PARTNERS]: string[];
};
