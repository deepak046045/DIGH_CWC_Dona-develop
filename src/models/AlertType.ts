export type AlertType = {
  isOpen: boolean;
  message: string;
  level: string;
};
