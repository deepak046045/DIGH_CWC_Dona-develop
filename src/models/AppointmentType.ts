import { AppointmentTypes } from '@/shared/enums/AppointmentTypes';

export type AppointmentType = {
  name: AppointmentTypes;
  count?: number;
};

export const APPOINTMENT_ICONS: { [key in AppointmentTypes]: string } = {
  [AppointmentTypes.INFUSION]: '/icons/iv-infusion-icon.svg',
  [AppointmentTypes.GENERAL_DOCTOR]:
    '/icons/general-practicioner-check-in-icon.svg',
  [AppointmentTypes.NEUROLOGIST]: '/icons/neurologist-icon.svg',
  [AppointmentTypes.SAFETY_MRI]: '/icons/safety-mri-icon.svg',
};
