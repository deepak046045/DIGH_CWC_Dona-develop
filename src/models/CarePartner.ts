export type CarePartner = {
  id: string;
  name: string;
  email: string;
};
