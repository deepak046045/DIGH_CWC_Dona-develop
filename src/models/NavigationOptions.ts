export type NavigationOption = {
  title: string;
  path: string;
};

export const NavigationOptions: Record<string, NavigationOption> = {
  TreatmentMap: {
    title: 'Treatment map',
    path: '/treatmentMap',
  },
  WhatToExpect: {
    title: 'What to expect',
    path: '/whatToExpect',
  },
  MyTreatmentPlan: {
    title: 'My treatment plan',
    path: '/myTreatmentPlan',
  },
  MyAppointments: {
    title: 'My appointments',
    path: '/myAppointments',
  },
  AddAppointment: {
    title: 'Add appointment',
    path: '/addAppointment',
  },
  MyCarePartners: {
    title: 'My care partners',
    path: '/myCarePartners',
  },
  MyProfile: {
    title: 'My profile',
    path: '/myProfile',
  },
};

export const gatedNavigationOptions: NavigationOption[] = [
  NavigationOptions.MyTreatmentPlan,
  NavigationOptions.MyAppointments,
  NavigationOptions.MyCarePartners,
  NavigationOptions.WhatToExpect,
];

export const ungatedNavigationOptions: NavigationOption[] = [
  NavigationOptions.TreatmentMap,
  NavigationOptions.WhatToExpect,
];
