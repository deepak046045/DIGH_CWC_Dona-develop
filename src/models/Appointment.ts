import { AppointmentType } from './AppointmentType';
import { CarePartner } from './CarePartner';

export type Appointment = {
  id: string;
  date: Date;
  address: string;
  type: AppointmentType;
  carePartners: CarePartner[];
};
