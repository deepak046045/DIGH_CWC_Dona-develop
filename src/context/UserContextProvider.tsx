'use client';

import React, {
  Dispatch,
  ReactNode,
  SetStateAction,
  createContext,
  useMemo,
  useState,
} from 'react';
import {
  NavigationOption,
  ungatedNavigationOptions,
} from '@/models/NavigationOptions';
import { Appointment } from '@/models/Appointment';
import { AlertType } from '@/models/AlertType';
import mockAppointments from '@/mocks/mockAppointments';
import { CarePartner } from '@/models/CarePartner';
import { mockCarePartners } from '@/mocks/mockCarePartners';

type UserContextDataType = {
  isLoggedIn: boolean;
  setIsLoggedIn: Dispatch<SetStateAction<boolean>>;
  activeView: NavigationOption;
  setActiveView: Dispatch<NavigationOption>;
  alert: AlertType;
  setAlert: Dispatch<SetStateAction<AlertType>>;
  appointments: Appointment[];
  setAppointments: Dispatch<SetStateAction<Appointment[]>>;
  carePartners: CarePartner[];
  setCarePartners: Dispatch<SetStateAction<CarePartner[]>>;
};

export const UserContext = createContext<UserContextDataType>({
  isLoggedIn: false,
  setIsLoggedIn: () => {},
  activeView: ungatedNavigationOptions[0],
  setActiveView: () => {},
  alert: { isOpen: false, message: '', level: '' },
  setAlert: () => {},
  appointments: mockAppointments,
  setAppointments: () => {},
  carePartners: mockCarePartners,
  setCarePartners: () => {},
});

type Props = {
  children: ReactNode;
};

export function UserContextProvider({ children }: Props) {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [activeView, setActiveView] = useState<NavigationOption>(
    ungatedNavigationOptions[0]
  );
  const [alert, setAlert] = useState<AlertType>({
    isOpen: false,
    message: '',
    level: '',
  });
  const [appointments, setAppointments] =
    useState<Appointment[]>(mockAppointments);
  const [carePartners, setCarePartners] =
    useState<CarePartner[]>(mockCarePartners);

  const values = useMemo(
    () => ({
      isLoggedIn,
      setIsLoggedIn,
      activeView,
      setActiveView,
      alert,
      setAlert,
      appointments,
      setAppointments,
      carePartners,
      setCarePartners,
    }),
    [isLoggedIn, activeView, alert, appointments, carePartners]
  );

  return <UserContext.Provider value={values}>{children}</UserContext.Provider>;
}
