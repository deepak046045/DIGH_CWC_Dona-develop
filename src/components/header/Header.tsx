'use client';

import Image from 'next/image';
import { Button, Divider } from '@mui/material';
import { LdsButton, LdsSticky } from '@elilillyco/ux-lds-react';
import React, { useState, useEffect, useContext } from 'react';
import { useRouter } from 'next/navigation';
import { UserContext } from '@/context/UserContextProvider';
import mockUserData from '@/mocks/mockUserData.json';
import HeaderMobile from '@/components/header/HeaderMobile';
import buttons from '@/scss/themes/buttons.module.scss';
import layout from '@/scss/themes/layout.module.scss';
import {
  NavigationOptions,
  NavigationOption,
  gatedNavigationOptions,
  ungatedNavigationOptions,
} from '@/models/NavigationOptions';
import { copyTexts } from '@/constants/copyTexts';
import headerSharedStyles from './HeaderShared.module.scss';
import headerStyles from './Header.module.scss';

export function Header() {
  const [isMobile, setIsMobile] = useState<boolean>(false);
  const { isLoggedIn, setIsLoggedIn, activeView, setActiveView } =
    useContext(UserContext);

  const router = useRouter();

  useEffect(() => {
    setIsMobile(window.innerWidth <= 767);
  }, []);

  useEffect(() => {
    const handleResize = () => {
      setIsMobile(window.innerWidth <= 767);
    };

    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  const signIn = (): void => {
    setIsLoggedIn(true);
    setActiveView(gatedNavigationOptions[0]);
  };

  const signOut = (): void => {
    setIsLoggedIn(false);
    setActiveView(ungatedNavigationOptions[0]);
    router.push('/');
  };

  const handleNavigation = (navigationOption: NavigationOption) => {
    setActiveView(navigationOption);
    router.push(navigationOption.path);
  };

  function buildNavigationButton(
    navigationOption: NavigationOption
  ): JSX.Element {
    return (
      <Button
        key={navigationOption.title}
        variant="contained"
        onClick={() => handleNavigation(navigationOption)}
        className={
          activeView === navigationOption
            ? headerStyles.navigationButtonActive
            : headerStyles.navigationButton
        }
      >
        {navigationOption.title}
      </Button>
    );
  }

  function buildNavigationButtonSelection(): JSX.Element {
    if (isLoggedIn) {
      return (
        <div className={`${headerStyles.grayBar} ${headerStyles.buttons}`}>
          {gatedNavigationOptions.map((navigationOption) =>
            buildNavigationButton(navigationOption)
          )}
        </div>
      );
    }

    return (
      <div className={`${headerStyles.grayBar} ${headerStyles.buttons}`}>
        {ungatedNavigationOptions.map((navigationOption) =>
          buildNavigationButton(navigationOption)
        )}
      </div>
    );
  }

  function buildSignInUpSection(): JSX.Element {
    if (isLoggedIn) {
      return (
        <div className={`${headerStyles.buttons}`}>
          <div className={headerSharedStyles.userWelcome}>
            Hi {mockUserData.firstName}
          </div>
          <Divider
            orientation="vertical"
            variant="middle"
            flexItem
            className={headerStyles.buttonDivider}
          />
          <Button
            onClick={() => handleNavigation(NavigationOptions.MyProfile)}
            className={
              activeView === NavigationOptions.MyProfile
                ? headerSharedStyles.myProfileButtonActive
                : headerSharedStyles.myProfileButton
            }
          >
            <Image
              src={
                activeView === NavigationOptions.MyProfile
                  ? '/icons/profile-active-icon.svg'
                  : '/icons/profile-icon.svg'
              }
              alt="ProfileIcon"
              width="35"
              height="35"
            />
            {NavigationOptions.MyProfile.title}
          </Button>
          <Divider
            orientation="vertical"
            variant="middle"
            flexItem
            className={headerStyles.buttonDivider}
          />
          <LdsButton
            onClick={signOut}
            data-testid="sign-out-button"
            classes={`${buttons.primary}`}
          >
            {copyTexts.Header.signOut}
          </LdsButton>
        </div>
      );
    }

    return (
      <div className={headerStyles.buttons}>
        <LdsButton
          onClick={signIn}
          data-testid="sign-in-button"
          classes={`${buttons.secondary}`}
        >
          {copyTexts.Header.signIn}
        </LdsButton>
        <Divider
          orientation="vertical"
          variant="middle"
          flexItem
          className={headerStyles.buttonDivider}
        />
        <LdsButton classes={`${buttons.primary}`}>
          {copyTexts.Header.signUp}
        </LdsButton>
      </div>
    );
  }

  return (
    <LdsSticky>
      {isMobile ? (
        <HeaderMobile
          isLoggedIn={isLoggedIn}
          signIn={signIn}
          signOut={signOut}
          activeView={activeView}
          handleNavigation={handleNavigation}
        />
      ) : (
        <div className={headerSharedStyles.background}>
          <div className={`${headerSharedStyles.whiteBar} ${layout.gutter}`}>
            <Image
              src="/images/kisunla-logo.svg"
              alt="Kisunla Logo"
              width={297}
              height={98}
              priority
            />
            {buildSignInUpSection()}
          </div>
          {buildNavigationButtonSelection()}
        </div>
      )}
    </LdsSticky>
  );
}
