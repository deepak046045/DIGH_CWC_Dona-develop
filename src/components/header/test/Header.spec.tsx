import { fireEvent, render, screen } from '@testing-library/react';
import { ungatedNavigationOptions } from '@/models/NavigationOptions';
import { UserContextProvider } from '@/context/UserContextProvider';
import React from 'react';
import { Header } from '../Header';
import '@testing-library/jest-dom';

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(() => ({ push: jest.fn() })),
}));
jest.mock('crypto', () => ({ randomUUID: () => '123456789' }));

describe('Header Component', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: 1000,
    });
  });

  it('should render desktop view', () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });

  describe('window resizing', () => {
    it('should render mobile view', () => {
      window.innerWidth = 500;
      const { container } = render(<Header />);
      expect(container).toMatchSnapshot();
    });

    it('should handle resize and show mobile view when window size changes', () => {
      const { getByTestId, queryByTestId } = render(<Header />);
      expect(queryByTestId('mobile-menu-button')).not.toBeInTheDocument();
      window.innerWidth = 500;
      fireEvent(window, new Event('resize'));
      expect(getByTestId('mobile-menu-button')).toBeInTheDocument();
    });
  });

  describe('when a navigation button is clicked', () => {
    it('should swap active status', () => {
      render(
        <UserContextProvider>
          <Header />
        </UserContextProvider>
      );
      const firstNavigationButton = screen.getByText(
        ungatedNavigationOptions[0].title
      );
      const secondNavigationButton = screen.getByText(
        ungatedNavigationOptions[1].title
      );

      expect(firstNavigationButton).toHaveClass('navigationButtonActive');
      expect(secondNavigationButton).toHaveClass('navigationButton');

      fireEvent.click(secondNavigationButton);

      expect(firstNavigationButton).toHaveClass('navigationButton');
      expect(secondNavigationButton).toHaveClass('navigationButtonActive');
    });
  });

  describe('when clicking the sign in button', () => {
    it('should sign user in', () => {
      const { queryByTestId } = render(
        <UserContextProvider>
          <Header />
        </UserContextProvider>
      );
      const signInButton = screen.getByText('Sign in');
      expect(queryByTestId('sign-out-button')).not.toBeInTheDocument();

      fireEvent.click(signInButton);

      expect(queryByTestId('sign-out-button')).toBeInTheDocument();
    });
  });

  describe('when clicking the sign out button', () => {
    it('should sign user out', () => {
      const { queryByTestId } = render(
        <UserContextProvider>
          <Header />
        </UserContextProvider>
      );
      const signInButton = screen.getByText('Sign in');
      fireEvent.click(signInButton);

      const signOutButton = screen.getByText('Sign out');
      expect(queryByTestId('sign-in-button')).not.toBeInTheDocument();

      fireEvent.click(signOutButton);

      expect(queryByTestId('sign-in-button')).toBeInTheDocument();
    });
  });

  describe('when clicking the my profile button', () => {
    it('should set profile to active state', () => {
      render(
        <UserContextProvider>
          <Header />
        </UserContextProvider>
      );
      const signInButton = screen.getByText('Sign in');
      fireEvent.click(signInButton);

      const myProfileButton = screen.getByText('My profile');

      expect(myProfileButton).not.toHaveClass('myProfileButtonActive');

      fireEvent.click(myProfileButton);

      expect(myProfileButton).toHaveClass('myProfileButtonActive');
    });
  });
});
