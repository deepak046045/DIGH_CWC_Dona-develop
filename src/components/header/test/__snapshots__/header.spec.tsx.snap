// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should render desktop view 1`] = `
<div>
  <div
    class="lds-sticky"
    style="top: 0px; position: sticky; z-index: 100;"
  >
    <div
      class="background"
    >
      <div
        class="whiteBar gutter"
      >
        <img
          alt="Kisunla Logo"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="98"
          src="/images/kisunla-logo.svg"
          style="color: transparent;"
          width="297"
        />
        <div
          class="buttons"
        >
          <button
            class="lds-button secondary"
            data-testid="sign-in-button"
            type="button"
          >
            Sign in
          </button>
          <hr
            class="MuiDivider-root MuiDivider-middle MuiDivider-vertical MuiDivider-flexItem buttonDivider css-16bghm4-MuiDivider-root"
          />
          <button
            class="lds-button primary"
            type="button"
          >
            Sign up
          </button>
        </div>
      </div>
      <div
        class="grayBar buttons"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium navigationButtonActive css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Treatment map
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium navigationButton css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          What to expect
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header Component window resizing should render mobile view 1`] = `
<div>
  <div
    class="lds-sticky"
    style="top: 0px; position: sticky; z-index: 100;"
  >
    <div
      class="background"
    >
      <div
        class="whiteBar"
      >
        <img
          alt="Kisunla Logo"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="59"
          src="/images/kisunla-logo.svg"
          style="color: transparent;"
          width="152"
        />
        <div
          class="topRightControls"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="mobile-menu-button"
            tabindex="0"
            type="button"
          >
            <img
              alt="MenuHamburgerIcon"
              data-nimg="1"
              decoding="async"
              height="44"
              loading="lazy"
              src="/icons/menu-hamburger-icon.svg"
              style="color: transparent;"
              width="31"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
